apollo-router/src/plugins/cache/entity.rs

```diff
@@ -500,10 +514,33 @@ impl EntityCache {
                 concurrent_requests: 10,
             })),
             invalidation,
+            subgraph_enums: Arc::new(get_subgraph_enums(&supergraph_schema)),
+            supergraph_schema,
         })
     }
 }
 
+/// Get the map of subgraph enum variant mapped with subgraph name
+fn get_subgraph_enums(supergraph_schema: &Valid<Schema>) -> HashMap<String, String> {
+    let mut subgraph_enums = HashMap::new();
+    if let Some(graph_enum) = supergraph_schema.get_enum("join__Graph") {
+        subgraph_enums.extend(graph_enum.values.iter().filter_map(
+            |(enum_name, enum_value_def)| {
+                let subgraph_name = enum_value_def
+                    .directives
+                    .get("join__graph")
+                    .and_then(|directive| directive.specified_argument_by_name("name"))
+                    .and_then(|arg| arg.as_str())?
+                    .to_string();
```

Comment:
Nit (feel free to ignore): I think this might work:
```suggestion
                    .get("join__graph")?
                    .specified_argument_by_name("name")?
                    .as_str()?
                    .to_string();
```